File: typescript/src/algokit.ts
// SPDX-License-Identifier: MIT
// Copyright (C) 2026 APlane Project LLC

import { decodeAddress } from "algosdk";
import type { SignerClient } from "./client.js";
import { SignerError } from "./errors.js";
import { bytesToHex, encodeLsigArgs, hexToBytes } from "./encoding.js";
import type {
  GroupSignResponse,
  KeyInfo,
  LsigArgs,
  SignRequest,
  SignOptions,
} from "./types.js";

export type AlgoKitTransaction = {
  sender?: { toString(): string } | string;
};

export type AlgoKitTransactionSigner = (
  txnGroup: AlgoKitTransaction[],
  indexesToSign: number[],
) => Promise<Uint8Array[]>;

export type AlgoKitAddress = {
  readonly publicKey: Uint8Array;
  toString(): string;
  equals?(other: unknown): boolean;
};

type RequestsSignerClient = Pick<SignerClient, "signRequests" | "listKeys">;

export type AlgoKitTransactionEncoder = (
  txn: AlgoKitTransaction,
) => Uint8Array | Promise<Uint8Array>;

type AlgoKitEncoderModule = {
  encodeTransaction?: AlgoKitTransactionEncoder;
};

export interface ApsignerAccountOptions {
  /** Native APlane signer client. */
  client: RequestsSignerClient;
  /** Sending address exposed to AlgoKit as addr. */
  address: string;
  /** Apsigner key address to authorize the transaction. Defaults to address. */
  authAddress?: string;
  /** Runtime LogicSig args to send with each signed transaction. */
  lsigArgs?: LsigArgs;
  /** Optional factory that returns a caller-owned request ID per sign call. */
  newRequestId?: () => string;
  /** Optional cancellation signal for the /sign request. */
  signal?: AbortSignal;
  /** Override transaction encoding, mostly useful for tests. */
  encodeTransaction?: AlgoKitTransactionEncoder;
  /** Override the object exposed as addr. Defaults to an Algorand-address-compatible object. */
  addressObject?: AlgoKitAddress;
}

export interface ApsignerAccount {
  readonly addr: AlgoKitAddress;
  readonly authAddress: string;
  readonly signer: AlgoKitTransactionSigner;
}

class AlgorandAddressLike implements AlgoKitAddress {
  readonly publicKey: Uint8Array;

  constructor(private readonly value: string) {
    this.publicKey = decodeAddress(value).publicKey;
  }

  toString(): string {
    return this.value;
  }

  equals(other: unknown): boolean {
    if (!other) {
      return false;
    }
    if (typeof other === "string") {
      return other === this.value;
    }
    if (typeof (other as { toString?: unknown }).toString === "function") {
      return (other as { toString(): string }).toString() === this.value;
    }
    return false;
  }
}

async function dynamicImport<T>(specifier: string): Promise<T> {
  const importer = new Function("specifier", "return import(specifier)") as (
    specifier: string,
  ) => Promise<T>;
  return importer(specifier);
}

async function loadAlgoKitEncoder(
  specifier: string,
): Promise<AlgoKitTransactionEncoder | undefined> {
  try {
    const mod = await dynamicImport<AlgoKitEncoderModule>(specifier);
    return typeof mod.encodeTransaction === "function" ? mod.encodeTransaction : undefined;
  } catch {
    return undefined;
  }
}

async function defaultEncodeAlgoKitTransaction(txn: AlgoKitTransaction): Promise<Uint8Array> {
  const encodeTransaction =
    await loadAlgoKitEncoder("@algorandfoundation/algokit-utils/transact")
    ?? await loadAlgoKitEncoder("@algorandfoundation/algokit-transact");
  if (!encodeTransaction) {
    throw new SignerError(
      "AlgoKit transaction encoder not found; install @algorandfoundation/algokit-utils@10 or pass encodeTransaction",
    );
  }
  return encodeTransaction(txn);
}

function txnSender(txn: AlgoKitTransaction): string {
  const sender = txn.sender;
  if (typeof sender === "string") {
    return sender;
  }
  if (sender && typeof sender.toString === "function") {
    return sender.toString();
  }
  throw new SignerError("AlgoKit transaction is missing sender");
}

/**
 * AlgoKit AddressWithTransactionSigner adapter backed by apsigner.
 *
 * The adapter connects AlgoKit clients to APlane's transaction signing functions.
 */
export class ApsignerAlgoKitAccount implements ApsignerAccount {
  readonly addr: AlgoKitAddress;
  readonly authAddress: string;
  readonly signer: AlgoKitTransactionSigner;

  private readonly client: RequestsSignerClient;
  private readonly lsigArgs?: LsigArgs;
  private readonly newRequestId?: () => string;
  private readonly signal?: AbortSignal;
  private readonly encodeTransaction: AlgoKitTransactionEncoder;

  constructor(options: ApsignerAccountOptions) {
    this.client = options.client;
    this.addr = options.addressObject ?? new AlgorandAddressLike(options.address);
    this.authAddress = options.authAddress ?? options.address;
    this.lsigArgs = options.lsigArgs;
    this.newRequestId = options.newRequestId;
    this.signal = options.signal;
    this.encodeTransaction = options.encodeTransaction ?? defaultEncodeAlgoKitTransaction;
    this.signer = async (txnGroup, indexesToSign) => this.sign(txnGroup, indexesToSign);
  }

  private nextRequestId(): string | undefined {
    return this.newRequestId?.();
  }

  private async sign(
    txnGroup: AlgoKitTransaction[],
    indexesToSign: number[],
  ): Promise<Uint8Array[]> {
    if (indexesToSign.length === 0) {
      return [];
    }

    const requests: SignRequest[] = [];
    for (const index of indexesToSign) {
      if (index < 0 || index >= txnGroup.length) {
        throw new SignerError(`index ${index} out of range for ${txnGroup.length} transactions`);
      }
      const txn = txnGroup[index];
      if (!txn) {
        throw new SignerError(`transaction is required at index ${index}`);
      }

      const req: SignRequest = {
        txn_bytes_hex: bytesToHex(await this.encodeTransaction(txn)),
        txn_sender: txnSender(txn),
        auth_address: this.authAddress,
      };
      if (this.lsigArgs) {
        req.lsig_args = encodeLsigArgs(this.lsigArgs);
      }
      requests.push(req);
    }

    const response = await this.client.signRequests(
      requests,
      {
        requestId: this.nextRequestId(),
        signal: this.signal,
      } satisfies SignOptions,
    );

    const signed = response.signed ?? [];
    if (signed.length < indexesToSign.length) {
      throw new SignerError(
        "apsigner returned fewer signed transactions than AlgoKit requested",
      );
    }

    return signed.map((hex) => hexToBytes(hex));
  }
}

export function createApsignerAccount(options: ApsignerAccountOptions): ApsignerAccount {
  return new ApsignerAlgoKitAccount(options);
}

export async function listApsignerAccounts(
  client: RequestsSignerClient,
  options: { refresh?: boolean } = {},
): Promise<ApsignerAccount[]> {
  const keys: KeyInfo[] = await client.listKeys(options.refresh ?? false);
  return keys.map((key) => createApsignerAccount({
    client,
    address: key.address,
    authAddress: key.address,
  }));
}

export type { GroupSignResponse };

File: python/aplanesdk/algokit.py
# SPDX-License-Identifier: MIT
# Copyright (C) 2026 APlane Project LLC

"""AlgoKit Utils adapter for apsigner-backed transaction signing."""

from collections.abc import Callable, Sequence
from secrets import token_hex
from threading import Lock
from typing import Any, Optional, Protocol

from .signer import CancelSignResponse, GroupSignResponse, KeyInfo, SignerError


class _GroupSignerClient(Protocol):
    def sign_requests(
        self,
        sign_entries: list[dict[str, Any]],
        *,
        request_id: Optional[str] = None,
    ) -> GroupSignResponse: ...

    def list_keys(self, refresh: bool = False) -> list[KeyInfo]: ...

    def cancel_sign_request(self, request_id: str) -> CancelSignResponse: ...


TransactionEncoder = Callable[[Any], bytes]
RequestIDFactory = Callable[[], str]


def _default_request_id() -> str:
    return f"sdk-{token_hex(16)}"


def _default_encode_transaction(txn: Any) -> bytes:
    try:
        from algokit_transact.codec.transaction import encode_transaction
    except ImportError as exc:
        raise SignerError(
            "algokit_transact is not importable; install algokit-utils >= 5.0.0b1 "
            "or pass an explicit encode_transaction"
        ) from exc
    return encode_transaction(txn)


def _txn_sender(txn: Any) -> str:
    sender = getattr(txn, "sender", None)
    if sender is None:
        raise SignerError("AlgoKit transaction is missing sender")
    return str(sender)


class ApsignerAccount:
    """
    AlgoKit AddressWithTransactionSigner adapter backed by apsigner.

    The adapter connects AlgoKit clients to APlane's transaction signing
    functions.
    """

    def __init__(
        self,
        client: _GroupSignerClient,
        address: str,
        *,
        auth_address: Optional[str] = None,
        lsig_args: Optional[dict[str, bytes]] = None,
        new_request_id: Optional[RequestIDFactory] = None,
        encode_transaction: Optional[TransactionEncoder] = None,
    ) -> None:
        self._client = client
        self._addr = address
        self.auth_address = auth_address or address
        self._lsig_args = lsig_args
        self._new_request_id = new_request_id or _default_request_id
        self._encode_transaction = encode_transaction or _default_encode_transaction
        self._current_request_id: Optional[str] = None
        self._lock = Lock()

    @property
    def addr(self) -> str:
        return self._addr

    @property
    def signer(self) -> Callable[[Sequence[Any], Sequence[int]], list[bytes]]:
        return self._sign

    def _begin_sign(self) -> str:
        request_id = self._new_request_id()
        with self._lock:
            if self._current_request_id is not None:
                raise RuntimeError("ApsignerAccount already has an in-flight signing request")
            self._current_request_id = request_id
        return request_id

    def _end_sign(self, request_id: str) -> None:
        with self._lock:
            if self._current_request_id == request_id:
                self._current_request_id = None

    def cancel(self) -> None:
        """
        Best-effort cancellation for the current synchronous AlgoKit sign call.

        AlgoKit's Python signer callback has no cancellation parameter, so
        applications that need user cancellation can call this from another
        thread while the signer is waiting for operator approval.
        """
        with self._lock:
            request_id = self._current_request_id
        if request_id is None:
            return
        try:
            self._client.cancel_sign_request(request_id)
        except (SignerError, OSError):
            # Best-effort: backend errors and transport failures are tolerated,
            # but programming errors (TypeError, AttributeError, ...) propagate.
            pass

    def _sign(self, txn_group: Sequence[Any], indexes_to_sign: Sequence[int]) -> list[bytes]:
        if not indexes_to_sign:
            return []

        request_id = self._begin_sign()
        try:
            requests: list[dict[str, Any]] = []
            for index in indexes_to_sign:
                if index < 0 or index >= len(txn_group):
                    raise SignerError(
                        f"index {index} out of range for {len(txn_group)} transactions"
                    )

                txn = txn_group[index]
                if txn is None:
                    raise SignerError(f"transaction is required at index {index}")

                req = {
                    "txn_bytes_hex": self._encode_transaction(txn).hex(),
                    "txn_sender": _txn_sender(txn),
                    "auth_address": self.auth_address,
                }
                if self._lsig_args:
                    req["lsig_args"] = {
                        name: value.hex()
                        for name, value in self._lsig_args.items()
                    }
                requests.append(req)

            result = self._client.sign_requests(requests, request_id=request_id)

            if len(result.signed) < len(indexes_to_sign):
                raise SignerError(
                    "apsigner returned fewer signed transactions than AlgoKit requested"
                )

            return [bytes.fromhex(item) for item in result.signed]
        finally:
            self._end_sign(request_id)


def create_apsigner_account(
    client: _GroupSignerClient,
    address: str,
    **kwargs: Any,
) -> ApsignerAccount:
    return ApsignerAccount(client, address, **kwargs)


def list_apsigner_accounts(
    client: _GroupSignerClient,
    *,
    refresh: bool = False,
) -> list[ApsignerAccount]:
    return [
        ApsignerAccount(client, key.address, auth_address=key.address)
        for key in client.list_keys(refresh=refresh)
    ]
